### PR TITLE
PP-2262 Unix last logins

### PR DESCRIPTION
**Pull request by joshmyers**
## What

See commits.

TL;DR This shouldn't report at all if instances were launched within the last 90 days.

## How

This is a hard one to test as it only reports on instances that are _older_ than 90 days by default. To test this I had to run aws_compliance, then flip the logic so it reports on instances _younger_ than 90 days.

* Run the following on a subset or all instances in an environment via Ansible, this will get run via [cron](https://github.com/alphagov/pay-chef/pull/255)

```
/usr/local/bin/pay-unix-last-login-then-report-to-s3 -b pay-govuk-unix-accounts-${AWS_ACCOUNT} -e ${DEPLOY_ENV}
```

This should generate some reports:

```
aws> s3 ls s3://pay-govuk-unix-accounts-dev/2017-07-04/dev-josh-25/
2017-07-04 13:51:13        614 dev-josh-25-adminusers-i-0e7316e7ba467647d__1499172671.json
2017-07-04 13:51:14        614 dev-josh-25-bastion-i-0d27a387c091c1e3a__1499172672.json
2017-07-04 13:51:15        614 dev-josh-25-carbon-relay-i-0157463fb1771af9e__1499172674.json
2017-07-04 13:47:37        614 dev-josh-25-cardid-i-095f2f97861c23c0e__1499172455.json
2017-07-04 13:51:12        614 dev-josh-25-cardid-i-095f2f97861c23c0e__1499172670.json
2017-07-04 13:51:16        614 dev-josh-25-connector-i-031553da5b65bf352__1499172674.json
2017-07-04 13:51:11        614 dev-josh-25-egress-proxy-i-02fb48e986225d756__1499172669.json
2017-07-04 13:51:16        614 dev-josh-25-notifications-i-0a2fb91bfec036d09__1499172674.json
2017-07-04 13:51:14        614 dev-josh-25-ntp-i-054ed46f50568f634__1499172673.json
2017-07-04 13:51:14        614 dev-josh-25-publicapi-i-01fd7fa315253b12f__1499172672.json
2017-07-04 13:51:12        614 dev-josh-25-publicauth-i-0b97d1587e90d24a7__1499172670.json
2017-07-04 13:51:14        614 dev-josh-25-selfservice-i-00d3654a7e124bf3c__1499172672.json
2017-07-04 13:51:13        614 dev-josh-25-vault-0__1499172671.json
2017-07-04 13:51:11        614 dev-josh-25-vault-1__1499172669.json
2017-07-04 13:51:12        614 dev-josh-25-www-i-0528592ad1e59b41d__1499172670.json
```

Run aws_compliance:

```
ONLY_SHOW_FAILED=true python aws_compliance.py     ⚡ dev
[
    {
        "Description": "Ensure S3 versioning is enabled on all buckets",
        "ScoredControl": false,
        "failReason": "Buckets found without versioning enabled",
        "Result": false,
        "Offenders": [
            "govuk-pay-dev-josh-repo"
        ],
        "ControlId": "s3_versioning_enabled"
    },
    {
        "Description": "Ensure S3 logging is enabled on all buckets",
        "ScoredControl": false,
        "failReason": "Buckets found without logging enabled",
        "Result": false,
        "Offenders": [
            "govuk-pay-dev-baz-22-repo",
            "govuk-pay-dev-josh-repo"
        ],
        "ControlId": "s3_logging_enabled"
    },
    {
        "Description": "Instances requiring a reboot, see /var/log/pay-reboots-required.log on instances",
        "ScoredControl": false,
        "failReason": "Instances found requiring reboots",
        "Result": false,
        "Offenders": [
            "dev-alan-14-ntp-i-0b8b5076983c0c8be",
            "dev-josh-25-adminusers-i-0e7316e7ba467647d",
            "dev-alan-14-selfservice-i-01d38a2f49d948178",
            "dev-josh-25-selfservice-i-00d3654a7e124bf3c",
            "dev-alan-14-cardid-i-08d99baa91990f156",
            "dev-alan-14-vault-0",
            "dev-alan-14-connector-i-0472051fce5e1a7ca",
            "dev-alan-14-vault-1",
            "dev-alan-14-notifications-i-0f432ae3e168ccdb7",
            "dev-josh-25-egress-proxy-i-02fb48e986225d756",
            "dev-alan-14-publicauth-i-0cc8cb6c9a67ed4c2",
            "dev-alan-14-www-i-0171f7f86b6cfe4c9",
            "dev-alan-14-publicapi-i-034da09f84f1aa4aa",
            "dev-alan-14-carbon-relay-i-0cd4cf12c3da7c897",
            "dev-alan-14-bastion-i-02bb042e51d2fdd1c",
            "dev-alan-14-egress-proxy-i-0e6e1debf15202289"
        ],
        "ControlId": "reboots_required"
    },
    {
        "Description": "Ensure multi-factor authentication (MFA) is enabled for all IAM users that have a console password",
        "ScoredControl": false,
        "failReason": "No MFA on users with password. ",
        "Result": false,
        "Offenders": [
            "arn:aws:iam::673337093959:user/dominic.jones"
        ],
        "ControlId": "mfa_on_password_enabled_iam"
    },
    {
        "Description": "Ensure credentials unused for 90 days or greater are disabled",
        "ScoredControl": false,
        "failReason": "Credentials unused > 90 days detected. ",
        "Result": false,
        "Offenders": [
            "arn:aws:iam::673337093959:user/alex.bishop:password",
            "arn:aws:iam::673337093959:user/danail.minchev:password",
            "arn:aws:iam::673337093959:user/danail.minchev:key1",
            "arn:aws:iam::673337093959:user/david.heath:password",
            "arn:aws:iam::673337093959:user/david.heath:key1",
            "arn:aws:iam::673337093959:user/david.king:password",
            "arn:aws:iam::673337093959:user/dominic.jones:password",
            "arn:aws:iam::673337093959:user/raul.garcia:password",
            "arn:aws:iam::673337093959:user/sandor.arpa:password",
            "arn:aws:iam::673337093959:user/sandor.arpa:key1"
        ],
        "ControlId": "unused_credentials"
    }
]
```

Flip logic so instances _younger_ than 90 days are included in the report:

```
ONLY_SHOW_FAILED=true python aws_compliance.py                                                                                                                             ⚡ dev
[
    {
        "Description": "Ensure S3 versioning is enabled on all buckets",
        "ScoredControl": false,
        "failReason": "Buckets found without versioning enabled",
        "Result": false,
        "Offenders": [
            "govuk-pay-dev-josh-repo"
        ],
        "ControlId": "s3_versioning_enabled"
    },
    {
        "Description": "Ensure S3 logging is enabled on all buckets",
        "ScoredControl": false,
        "failReason": "Buckets found without logging enabled",
        "Result": false,
        "Offenders": [
            "govuk-pay-dev-baz-22-repo",
            "govuk-pay-dev-josh-repo"
        ],
        "ControlId": "s3_logging_enabled"
    },
    {
        "Description": "Instances requiring a reboot, see /var/log/pay-reboots-required.log on instances",
        "ScoredControl": false,
        "failReason": "Instances found requiring reboots",
        "Result": false,
        "Offenders": [
            "dev-alan-14-ntp-i-0b8b5076983c0c8be",
            "dev-josh-25-adminusers-i-0e7316e7ba467647d",
            "dev-alan-14-selfservice-i-01d38a2f49d948178",
            "dev-josh-25-selfservice-i-00d3654a7e124bf3c",
            "dev-alan-14-cardid-i-08d99baa91990f156",
            "dev-alan-14-vault-0",
            "dev-alan-14-connector-i-0472051fce5e1a7ca",
            "dev-alan-14-vault-1",
            "dev-alan-14-notifications-i-0f432ae3e168ccdb7",
            "dev-josh-25-egress-proxy-i-02fb48e986225d756",
            "dev-alan-14-publicauth-i-0cc8cb6c9a67ed4c2",
            "dev-alan-14-www-i-0171f7f86b6cfe4c9",
            "dev-alan-14-publicapi-i-034da09f84f1aa4aa",
            "dev-alan-14-carbon-relay-i-0cd4cf12c3da7c897",
            "dev-alan-14-bastion-i-02bb042e51d2fdd1c",
            "dev-alan-14-egress-proxy-i-0e6e1debf15202289"
        ],
        "ControlId": "reboots_required"
    },
    {
        "Description": "Ensure multi-factor authentication (MFA) is enabled for all IAM users that have a console password",
        "ScoredControl": false,
        "failReason": "No MFA on users with password. ",
        "Result": false,
        "Offenders": [
            "arn:aws:iam::673337093959:user/dominic.jones"
        ],
        "ControlId": "mfa_on_password_enabled_iam"
    },
    {
        "Description": "Ensure credentials unused for 90 days or greater are disabled",
        "ScoredControl": false,
        "failReason": "Credentials unused > 90 days detected. ",
        "Result": false,
        "Offenders": [
            "arn:aws:iam::673337093959:user/alex.bishop:password",
            "arn:aws:iam::673337093959:user/danail.minchev:password",
            "arn:aws:iam::673337093959:user/danail.minchev:key1",
            "arn:aws:iam::673337093959:user/david.heath:password",
            "arn:aws:iam::673337093959:user/david.heath:key1",
            "arn:aws:iam::673337093959:user/david.king:password",
            "arn:aws:iam::673337093959:user/dominic.jones:password",
            "arn:aws:iam::673337093959:user/raul.garcia:password",
            "arn:aws:iam::673337093959:user/sandor.arpa:password",
            "arn:aws:iam::673337093959:user/sandor.arpa:key1"
        ],
        "ControlId": "unused_credentials"
    },
    {
        "Description": "Unix account last login older than 90 days",
        "ScoredControl": false,
        "failReason": "Unix accounts found with last login over 90 days ago",
        "Result": false,
        "Offenders": {
            "dev-josh-25-selfservice-i-00d3654a7e124bf3c": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-connector-i-031553da5b65bf352": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-www-i-0528592ad1e59b41d": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-ntp-i-054ed46f50568f634": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-carbon-relay-i-0157463fb1771af9e": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-adminusers-i-0e7316e7ba467647d": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-cardid-i-095f2f97861c23c0e": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-publicapi-i-01fd7fa315253b12f": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-vault-1": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-egress-proxy-i-02fb48e986225d756": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-publicauth-i-0b97d1587e90d24a7": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-notifications-i-0a2fb91bfec036d09": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-vault-0": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ],
            "dev-josh-25-bastion-i-0d27a387c091c1e3a": [
                "markjones",
                "danailminchev",
                "willpearson",
                "kasunkumara",
                "christophekalista",
                "raulgarcia",
                "alanlapthorn",
                "davidking",
                "barriebremner",
                "yvanjanssens",
                "alexbishop",
                "sandorarpa",
                "davidheath",
                "danieleocchipinti",
                "dominicjones"
            ]
        },
        "ControlId": "unix_account_last_login_reports"
    }
]
```

### CODE REVIEW

**Review by lapthorn — APPROVED** [review]
Looks good to me,

**Comment by lapthorn** [comment]
If it becomes too noisy - then we can deal with the noise in 90 days....
